go_lib/deckhouse-config/validator.go

```diff
@@ -46,74 +46,139 @@ type ValuesValidator interface {
 }
 
 type ValidationResult struct {
-	IsConverted    bool
-	Settings       map[string]interface{}
-	Version        int
-	VersionWarning string
+	IsConverted bool
+	Settings    map[string]interface{}
+	Version     int
+
+	// true if spec.settings can be converted and validated.
+	HasVersionedSettings bool
+
+	ValidateCRWarning string
+	ValidateCRError   string
+	ConversionError   string
+	ValidationError   string
+}
+
+func (v ValidationResult) HasError() bool {
+	return v.ConversionError != "" || v.ValidateCRError != "" || v.ValidationError != ""
 }
 
-// Validate checks version value if spec.settings value is not empty,
-// then converts spec.settings to the latest version and finally validates
-// spec.settings using OpenAPI schema defined in related config-values.yaml file.
-func (c *ConfigValidator) Validate(cfg *d8cfg_v1alpha1.ModuleConfig) (ValidationResult, error) {
+func (v ValidationResult) Error() string {
+	errs := make([]string, 0)
+	if v.ValidateCRError != "" {
+		errs = append(errs, v.ValidateCRError)
+	}
+	if v.ConversionError != "" {
+		errs = append(errs, v.ConversionError)
+	}
+	if v.ValidationError != "" {
+		errs = append(errs, v.ValidationError)
+	}
+	return strings.Join(errs, ", ")
+}
+
+// ValidateCR checks if ModuleConfig resource is well-formed.
+func (c *ConfigValidator) ValidateCR(cfg *d8cfg_v1alpha1.ModuleConfig) ValidationResult {
```

Comment:
Let's call this method `ValidateModuleConfig` ?

Reply:
Nit: Instead of `ValidationResult`, it could return `error` that `ValidationResult` generates. And then Everything else can be deduced without `ValidationResult`, for example non-nil spec.settings.

Reply:
Discussed. Decision is to make refactoring later.
Use error wrapping and use v1alpha1.ModuleConfig object instead of additional type.